compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELogicalObjFifoSplittingUtils.cpp

```diff
@@ -427,4 +476,386 @@ LogicalResult splitLogicalObjectFifos(
+    SmallVector<OpFoldResult> &newStrides, SmallVector<int64_t> &splitDims,
+    SmallVector<int64_t> &nonSplitDims) {
+  if (offsetsA.empty() && offsetsB.empty()) return success();
+
+  int64_t newSize = 1;
+  for (auto iter : llvm::enumerate(llvm::zip(offsetsA, offsetsB))) {
+    if (iter.index() < splitDims.size()) continue;
+    const OpFoldResult &offsetA = std::get<0>(iter.value());
+    const OpFoldResult &offsetB = std::get<1>(iter.value());
+    if (offsetA != offsetB) {
+      // Need to check the difference in bias here.
+      int64_t biasA = fetchOffsetBias(offsetA);
+      int64_t biasB = fetchOffsetBias(offsetB);
+      std::optional<int64_t> sizeA = getConstantIntValue(sizesA[iter.index()]);
+      assert(sizeA && "expected a constant integer value for size");
+      assert((sizeA == biasB - biasA) &&
+             "L3->L2 pair cannot be combined because offset is not contiguous");
+      newSize++;
+    }
+  }
+  newSizes[splitDims.size() - 1] = rewriter.getI64IntegerAttr(newSize);
+  return success();
+}
+
+static FailureOr<LogicalObjectFifoFromMemrefOp> combineL3ToL2Pair(
+    IRRewriter &rewriter, DmaCpyNdOp dmaOpA, DmaCpyNdOp dmaOpB,
+    SmallVector<int64_t> &splitDims, SmallVector<int64_t> &nonSplitDims) {
+  OpBuilder::InsertionGuard guard(rewriter);
+  SmallVector<OpFoldResult> sourceOffsetsA = dmaOpA.getSourceMixedOffsets();
+  SmallVector<OpFoldResult> sourceSizesA = dmaOpA.getSourceMixedSizes();
+  SmallVector<OpFoldResult> sourceStridesA = dmaOpA.getSourceMixedStrides();
+  SmallVector<OpFoldResult> sourceOffsetsB = dmaOpB.getSourceMixedOffsets();
+  SmallVector<OpFoldResult> sourceSizesB = dmaOpB.getSourceMixedSizes();
+  SmallVector<OpFoldResult> sourceStridesB = dmaOpB.getSourceMixedStrides();
+
+  SmallVector<OpFoldResult> targetOffsetsA = dmaOpA.getTargetMixedOffsets();
+  SmallVector<OpFoldResult> targetSizesA = dmaOpA.getTargetMixedSizes();
+  SmallVector<OpFoldResult> targetStridesA = dmaOpA.getTargetMixedStrides();
+  SmallVector<OpFoldResult> targetOffsetsB = dmaOpB.getTargetMixedOffsets();
+  SmallVector<OpFoldResult> targetSizesB = dmaOpB.getTargetMixedSizes();
+  SmallVector<OpFoldResult> targetStridesB = dmaOpB.getTargetMixedStrides();
+
+  SmallVector<OpFoldResult> newSourceOffsets = sourceOffsetsA;
+  SmallVector<OpFoldResult> newSourceSizes = sourceSizesA;
+  SmallVector<OpFoldResult> newSourceStrides = sourceStridesA;
+  if (failed(combineL3ToL2AccessPatterns(
+          rewriter, sourceOffsetsA, sourceSizesA, sourceStridesA,
+          sourceOffsetsB, sourceSizesB, sourceStridesB, newSourceOffsets,
+          newSourceSizes, newSourceStrides, splitDims, nonSplitDims))) {
+    return failure();
+  }
+
+  SmallVector<OpFoldResult> newTargetOffsets = targetOffsetsA;
+  SmallVector<OpFoldResult> newTargetSizes = newSourceSizes;
+  SmallVector<OpFoldResult> newTargetStrides = targetStridesA;
+  // Now we need to create a new L2 buffer based on `newTargetSizes`.
```

Comment:
I might be missing it, but you probably need to check somewhere that the target access patterns of A and B are the same?